pkg/ingester-rf1/objstore/storage.go

```diff
@@ -0,0 +1,132 @@
+package objstore
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"sort"
+
+	"github.com/prometheus/common/model"
+
+	"github.com/grafana/loki/v3/pkg/storage"
+	"github.com/grafana/loki/v3/pkg/storage/chunk/client"
+	"github.com/grafana/loki/v3/pkg/storage/config"
+)
+
+type Multi struct {
+	stores        []*storeEntry
+	storageConfig storage.Config
+}
+
+type storeEntry struct {
+	start        model.Time
+	cfg          config.PeriodConfig
+	objectClient client.ObjectClient
+}
+
+var _ client.ObjectClient = (*Multi)(nil)
+
+func New(
+	periodicConfigs []config.PeriodConfig,
+	storageConfig storage.Config,
+	clientMetrics storage.ClientMetrics,
+) (*Multi, error) {
+	store := &Multi{
+		storageConfig: storageConfig,
+	}
+	// sort by From time
+	sort.Slice(periodicConfigs, func(i, j int) bool {
+		return periodicConfigs[i].From.Time.Before(periodicConfigs[i].From.Time)
+	})
+	for _, periodicConfig := range periodicConfigs {
+		objectClient, err := storage.NewObjectClient(periodicConfig.ObjectType, storageConfig, clientMetrics)
+		if err != nil {
+			return nil, fmt.Errorf("creating object client for period %s: %w ", periodicConfig.From, err)
+		}
+		store.stores = append(store.stores, &storeEntry{
+			start:        periodicConfig.From.Time,
+			cfg:          periodicConfig,
+			objectClient: objectClient,
+		})
+	}
+	return store, nil
+}
+
+func (m *Multi) GetStoreFor(ts model.Time) (client.ObjectClient, error) {
+	// find the schema with the lowest start _after_ tm
+	j := sort.Search(len(m.stores), func(j int) bool {
+		return m.stores[j].start > ts
+	})
+
+	// reduce it by 1 because we want a schema with start <= tm
+	j--
+
+	if 0 <= j && j < len(m.stores) {
+		return m.stores[j].objectClient, nil
+	}
```

Comment:
nit: if the first schema is after the `ts`, sort.Search() returns `0` , after this we decrease it and it will be `-1` that will fail this code.
but it's possible only in case the user misconfigured his schema_config.